### DISCUSSION TITLE
Pocketbase as Framework, respect API rules

### DISCUSSION THREAD

**Discussion by ccbiozhaw** [pocketbase/pocketbase#2910] [Q&A] [answered]
Hello, I am a huge fan of pocketbase, you are doing gods work. I want to extend my Application with additional search features, and decided on mirroring my Data in Meilisearch ( i asked about this yesterday ). I can get this working, but indexing can take time and i do not want to return stale data. 

I want to create an endpoint to search MS, return documents, take the ID, then refetch the data from the sqlite database. It might make searching slower, but the data I return will always be current. This will also make it easier to switch to different searching solutions.

Is it possible to respect the defined API filters when fetching data with app.Dao/app.dbx?
when i use 
```
		records, err := app.Dao().FindRecordsByIds("articles", []string{"a1ntf15thfywizz", "j3uhv63wi1phxdm"})
```

it will return the records, even though the user is not registered.

I think the alternative would be to return the Ids to the user in a first response, then use them to fetch the records again.

Thank you

**Comment by ganigeorgiev** [comment, marked as answer]
> Is it possible to respect the defined API filters when fetching data with app.Dao/app.dbx?

You can, but at the moment is not very pretty since you'll need to extract the current request data and to parse the filter rule that you want to check (in your case I guess the View rule should be used similar to the expand since you have the ids). It could look something like:

```go
e.Router.GET("/example", func(c echo.Context) error {
	// fetch the record ids from Meilisearch...
	ids := []string{"id1", "id2", "id3"}

	collection, err := app.Dao().FindCollectionByNameOrId("articles")
	if err != nil {
		return apis.NewBadRequestError("failed to fetch articles collection", err)
	}

	query := app.Dao().RecordQuery(collection).AndWhere(dbx.HashExp{
		// convert to list of any for IN expression
		(collection.Name + ".id"): list.ToInterfaceSlice(ids),
	})

	requestData := apis.RequestData(c)

	rule := collection.ViewRule

	if rule == nil && requestData.Admin == nil {
		return apis.NewForbiddenError("only admins can access this action", nil)
	}

	if rule != nil && *rule != "" {
		resolver := resolvers.NewRecordFieldResolver(
			app.Dao(),
			collection,
			requestData,
			// hidden fields are searchable only by admins
			requestData.Admin != nil,
		)
		expr, err := search.FilterData(*rule).BuildExpr(resolver)
		if err != nil {
			return apis.NewBadRequestError("failed to resolve api rule", err)
		}
		// attach any extra joins generated from the parsing
		// (note: this may change in the future and returned as BuildExpr result)
		resolver.UpdateQuery(query)
		// attach the parsed view rule conditions
		query.AndWhere(expr)
	}

	var records []*models.Record
	if err := query.All(&records); err != nil {
		return apis.NewBadRequestError("failed to fetch the searched records", err)
	}

	// allow client-side expand support
	apis.EnrichRecords(c, app.Dao(), records)

	return c.JSON(http.StatusOK, records)
})
```

**Reply by ccbiozhaw** [reply]
hmm I am getting the ``` apis.NewBadRequestError("Failed to resolve api rule.",", err)```

the only import i couldnt resolve is ```list.ToInterfaceSlice``` 
the implementation i found was this, which i assume is what you were doing:

```
func ToInterfaceSlice(slice []string) []interface{} {
	result := make([]interface{}, len(slice))
	for i, v := range slice {
		result[i] = v
	}
	return result
}
```

**Reply by ccbiozhaw** [reply]
if the collection is admin only i get a runtime error here:
```
			expr, err := search.FilterData(*collection.ViewRule).BuildExpr(resolver)
```

``` [PANIC RECOVER] runtime error: invalid memory address or nil pointer dereference goroutine 140 [running]:```

**Reply by ganigeorgiev** [reply]
> the only import i couldnt resolve is list.ToInterfaceSlice

gopls usually will autocomplete it for you, but the import path should be `github.com/pocketbase/pocketbase/tools/list`

> if the collection is admin only i get a runtime error here:

Ah, yes, If the rule is "Admin only" then `collection.ViewRule` will be nil, in this case you can ignore the resolver, aka. wrap it in a `if` (I'll update the above example shortly).

**Reply by ganigeorgiev** [reply]
_The example should resolve correctly now._

**Reply by ccbiozhaw** [reply]
for me it still fails here:

```
		if err := query.All(records); err != nil {
			return apis.NewBadRequestError("failed to fetch the searched records", err)
		}
```

pocketbase also prints an error to the console:

```2023/07/13 15:02:38 Invalid variable type: must be a pointer```

**Reply by ganigeorgiev** [reply]
There is no `type` symbol in the above example. Could you provide the code you are running?

**Reply by ccbiozhaw** [reply]
i think like this is works:
`if err := query.All(&records); err != nil {`
a pointer to the list records

**Reply by ganigeorgiev** [reply]
Ah, yes, sorry. The scan value must be a pointer.

**Reply by ccbiozhaw** [reply]
thank you so much:)

**Comment by hatemjaber** [comment]
I'm just asking out of curiosity... why an endpoint? Why not using the events API and set them for after add, after update, and after delete events instead? I think you can use the http API to make a request to MeiliSearch when the different events fire off.
